src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn set_mode(&self,
+        channel: PcntChannel,
+        pos_mode: PcntCountMode,
+        neg_mode: PcntCountMode,
+        hctrl_mode: PcntControlMode,
+        lctrl_mode: PcntControlMode) -> Result<(), EspError> {
+        unsafe {
+            esp!(pcnt_set_mode(self.unit, channel.into(), pos_mode.into(), neg_mode.into(), hctrl_mode.into(), lctrl_mode.into()))
+        }
+    }
+
+    #[doc = " @brief Add ISR handler for specified unit."]
+    #[doc = ""]
+    #[doc = " This ISR handler will be called from an ISR. So there is a stack"]
+    #[doc = " size limit (configurable as \"ISR stack size\" in menuconfig). This"]
+    #[doc = " limit is smaller compared to a global PCNT interrupt handler due"]
+    #[doc = " to the additional level of indirection."]
+    #[doc = ""]
+    #[doc = " @param callback Interrupt handler function."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub unsafe fn subscribe(&self, callback: impl FnMut(u32) + Send + 'static) -> Result<(), EspError> {
+        enable_isr_service()?;
+
+        //self.unsubscribe();
+        let callback: Box<dyn FnMut(u32) + 'static> = Box::new(callback);
+        ISR_HANDLERS[self.unit as usize] = Some(callback);
+        esp!(esp_idf_sys::pcnt_isr_handler_add(
+            self.unit,
+            Some(Self::handle_isr),
+            self.unit as *mut esp_idf_sys::c_types::c_void,
+        ))?;
+        Ok(())
+    }
+
+    #[doc = " @brief Remove ISR handler for specified unit."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn unsubscribe(&self) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_idf_sys::pcnt_isr_handler_remove(self.unit))?;
+            ISR_HANDLERS[self.unit as usize] = None;
+        }
+        Ok(())
+    }
+
+    unsafe extern "C" fn handle_isr(data: *mut esp_idf_sys::c_types::c_void) {
+        let unit = data as pcnt_unit_t;
+        if let Some(f) = &mut ISR_HANDLERS[unit as usize] {
+            let mut value = 0u32;
+            esp!(esp_idf_sys::pcnt_get_event_status(
+                unit,
+                &mut value as *mut u32
+            )).expect("failed to fetch event status!");
+            f(value);
+        }
+    }
+}
+
+impl Drop for Pcnt {
+    fn drop(&mut self) {
+        let _ = self.counter_pause();
+        let _ = self.intr_disable();
+        unsafe {ISR_HANDLERS[self.unit as usize] = None};
+        unit_deallocate(self.unit)
+    }
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static ISR_SERVICE_ENABLED: core::sync::atomic::AtomicBool =
+    core::sync::atomic::AtomicBool::new(false);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static PCNT_CS: crate::task::CriticalSection = crate::task::CriticalSection::new();
+
+fn enable_isr_service() -> Result<(), EspError> {
```

Comment:
Unnecessary when you have not enabled the `alloc` feature, so clippy will yell at you that this func is unused.